include/ginkgo/core/base/reordered.hpp

```diff
@@ -0,0 +1,244 @@
+            col_scaling_->rapply(system_matrix_.get(), system_matrix_.get());
+        }
+
+        // If a reordering factory is provided, generate the reordering and
+        // permute the system matrix accordingly.
+        if (parameters_.reordering) {
+            auto reordering = parameters_.reordering->generate(system_matrix_);
+            auto perm = reordering->get_permutation();
+            permutation_ =
+                clone(exec, as<matrix::Permutation<index_type>>(perm));
+            auto permutation_array =
+                make_array_view(exec, permutation_->get_size()[0],
+                                permutation_->get_permutation());
+            system_matrix_ = as<Permutable<index_type>>(system_matrix_)
+                                 ->permute(&permutation_array);
+        }
+
+        // Generate the inner operator with the scaled and reordered system
+        // matrix. If none is provided, use the Identity.
+        if (parameters_.inner_operator) {
+            inner_operator_ =
+                parameters_.inner_operator->generate(system_matrix_);
+        } else {
+            inner_operator_ = gko::matrix::Identity<value_type>::create(exec);
+        }
+    }
+
+    void apply_impl(const LinOp* b, LinOp* x) const override;
+
+    void apply_impl(const LinOp* alpha, const LinOp* b, const LinOp* beta,
+                    LinOp* x) const override;
+
+    /**
+     * Prepares the intermediate right hand side, solution and intermediate
+     * vectors for the inner operator by creating them and making sure they
+     * have the same sizes as `b` and `x`.
+     *
+     * @param b  Right hand side for the overall operator. Will be scaled and
+     * reordered and then serve as the right hand side for the inner operator.
+     * @param x  Solution vector and initial guess for the overall operator. In
+     * case the inner operator uses an initial guess, will be scaled and
+     * permuted accordingly.
+     */
+    void set_cache_to(const LinOp* b, const LinOp* x) const
+    {
+        if (cache_.inner_b == nullptr) {
+            const auto size = b->get_size();
+            cache_.inner_b =
+                matrix::Dense<value_type>::create(this->get_executor(), size);
+            cache_.inner_x =
+                matrix::Dense<value_type>::create(this->get_executor(), size);
+            cache_.intermediate =
+                matrix::Dense<value_type>::create(this->get_executor(), size);
+        }
+        cache_.inner_b->copy_from(b);
+        if (inner_operator_->apply_uses_initial_guess()) {
+            cache_.inner_x->copy_from(x);
+        }
+    }
+
+private:
+    std::shared_ptr<LinOp> system_matrix_{};
+    std::shared_ptr<const LinOp> inner_operator_{};
+    std::shared_ptr<const matrix::Diagonal<value_type>> row_scaling_{};
+    std::shared_ptr<const matrix::Diagonal<value_type>> col_scaling_{};
+    std::shared_ptr<matrix::Permutation<index_type>> permutation_{};
```

Comment:
Same here: if you want it to work for all reordering options, it should be a plain `LinOp` or a new data format (it should match the `get_permutation()` return type of your `ReorderingBase`).